File: common/src/main/java/com/dbx/agent/ConfiguredJdbcAgent.java
package com.dbx.agent;

import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Types;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public abstract class ConfiguredJdbcAgent implements DatabaseAgent {
    private final JdbcAgentProfile profile;
    private Connection connection;
    private String configuredDatabase = "";

    protected ConfiguredJdbcAgent(JdbcAgentProfile profile) {
        this.profile = profile;
    }

    public JdbcAgentProfile getProfile() {
        return profile;
    }

    @Override
    public Connection getConnection() {
        return connection;
    }

    @Override
    public void connect(ConnectParams params) {
        uncheckedVoid(() -> {
            Class.forName(profile.getDriverClass());
            connection = DriverManager.getConnection(profile.buildUrl(params), params.getUsername(), params.getPassword());
            configuredDatabase = params.getDatabase();
        });
    }

    @Override
    public boolean testConnection(ConnectParams params) {
        return unchecked(() -> {
            Class.forName(profile.getDriverClass());
            try (Connection conn = DriverManager.getConnection(profile.buildUrl(params), params.getUsername(), params.getPassword())) {
                return conn.isValid(5);
            }
        });
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return unchecked(() -> {
            Connection conn = requireConnection();
            Set<String> names = new LinkedHashSet<>();
            try {
                try (ResultSet rs = conn.getMetaData().getCatalogs()) {
                    while (rs.next()) {
                        addNonBlank(names, rs.getString("TABLE_CAT"));
                    }
                }
            } catch (Exception ignored) {
            }
            addNonBlank(names, configuredDatabase);
            addNonBlank(names, conn.getCatalog());

            List<DatabaseInfo> result = new ArrayList<>();
            for (String name : names) {
                result.add(new DatabaseInfo(name));
            }
            return result;
        });
    }

    @Override
    public List<String> listSchemas() {
        return unchecked(() -> {
            Connection conn = requireConnection();
            Set<String> names = new LinkedHashSet<>();
            DatabaseMetaData meta = conn.getMetaData();
            try {
                try (ResultSet rs = meta.getSchemas(null, null)) {
                    while (rs.next()) {
                        addNonBlank(names, rs.getString("TABLE_SCHEM"));
                    }
                }
            } catch (Exception first) {
                try (ResultSet rs = meta.getSchemas()) {
                    while (rs.next()) {
                        addNonBlank(names, rs.getString("TABLE_SCHEM"));
                    }
                }
            }
            addNonBlank(names, conn.getSchema());

            List<String> result = new ArrayList<>();
            for (String name : names) {
                if (!profile.getExcludedSchemas().contains(name.toUpperCase(Locale.ROOT))) {
                    result.add(name);
                }
            }
            Collections.sort(result);
            return result;
        });
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            DatabaseMetaData meta = requireConnection().getMetaData();
            List<TableInfo> result = new ArrayList<>();
            appendTables(result, meta, null, schema);
            if (result.isEmpty() && !configuredDatabase.trim().isEmpty()) {
                appendTables(result, meta, configuredDatabase, schema);
            }
            result.sort(Comparator.comparing(TableInfo::getName));
            return result;
        });
    }

    private void appendTables(List<TableInfo> result, DatabaseMetaData meta, String catalog, String schema) throws Exception {
        String[] tableTypes = profile.getTableTypes().toArray(new String[profile.getTableTypes().size()]);
        try (ResultSet rs = meta.getTables(catalog, blankToNull(schema), "%", tableTypes)) {
            while (rs.next()) {
                result.add(new TableInfo(
                    rs.getString("TABLE_NAME"),
                    normalizeTableType(rs.getString("TABLE_TYPE")),
                    rs.getString("REMARKS")
                ));
            }
        }
    }

    @Override
    public List<ObjectInfo> listObjects(String schema) {
        List<ObjectInfo> result = new ArrayList<>();
        for (TableInfo table : listTables(schema)) {
            result.add(new ObjectInfo(table.getName(), table.getTable_type(), schema, table.getComment()));
        }
        return result;
    }

    @Override
    public ObjectSource getObjectSource(String schema, String name, String objectType) {
        throw new UnsupportedOperationException("Object source is not supported");
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            DatabaseMetaData meta = requireConnection().getMetaData();
            Set<String> primaryKeys = primaryKeys(meta, null, schema, table);
            List<ColumnInfo> result = new ArrayList<>();
            appendColumns(result, meta, null, schema, table, primaryKeys);
            if (result.isEmpty() && !configuredDatabase.trim().isEmpty()) {
                Set<String> fallbackPrimaryKeys = primaryKeys(meta, configuredDatabase, schema, table);
                appendColumns(result, meta, configuredDatabase, schema, table, fallbackPrimaryKeys);
            }
            return result;
        });
    }

    private Set<String> primaryKeys(DatabaseMetaData meta, String catalog, String schema, String table) {
        Set<String> keys = new LinkedHashSet<>();
        try (ResultSet rs = meta.getPrimaryKeys(catalog, blankToNull(schema), table)) {
            while (rs.next()) {
                String name = rs.getString("COLUMN_NAME");
                if (name != null) {
                    keys.add(name);
                }
            }
        } catch (Exception ignored) {
        }
        return keys;
    }

    private void appendColumns(
        List<ColumnInfo> result,
        DatabaseMetaData meta,
        String catalog,
        String schema,
        String table,
        Set<String> primaryKeys
    ) throws Exception {
        try (ResultSet rs = meta.getColumns(catalog, blankToNull(schema), table, "%")) {
            while (rs.next()) {
                String name = rs.getString("COLUMN_NAME");
                result.add(new ColumnInfo(
                    name,
                    rs.getString("TYPE_NAME"),
                    rs.getInt("NULLABLE") != DatabaseMetaData.columnNoNulls,
                    rs.getString("COLUMN_DEF"),
                    primaryKeys.contains(name),
                    null,
                    rs.getString("REMARKS"),
                    intOrNull(rs, "COLUMN_SIZE"),
                    intOrNull(rs, "DECIMAL_DIGITS"),
                    characterLength(rs)
                ));
            }
        }
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return unchecked(() -> {
            DatabaseMetaData meta = requireConnection().getMetaData();
            Map<String, List<IndexColumn>> indexes = new LinkedHashMap<>();
            Map<String, Boolean> unique = new LinkedHashMap<>();
            try (ResultSet rs = meta.getIndexInfo(null, blankToNull(schema), table, false, false)) {
                while (rs.next()) {
                    String name = rs.getString("INDEX_NAME");
                    String column = rs.getString("COLUMN_NAME");
                    if (name == null || column == null) {
                        continue;
                    }
                    indexes.computeIfAbsent(name, ignored -> new ArrayList<>()).add(
                        new IndexColumn(rs.getShort("ORDINAL_POSITION"), column)
                    );
                    unique.put(name, !rs.getBoolean("NON_UNIQUE"));
                }
            }

            List<IndexInfo> result = new ArrayList<>();
            for (Map.Entry<String, List<IndexColumn>> entry : indexes.entrySet()) {
                List<IndexColumn> orderedColumns = entry.getValue();
                orderedColumns.sort(Comparator.comparingInt(IndexColumn::getOrdinal));
                List<String> columns = new ArrayList<>();
                for (IndexColumn column : orderedColumns) {
                    columns.add(column.getName());
                }
                String name = entry.getKey();
                result.add(new IndexInfo(
                    name,
                    columns,
                    Boolean.TRUE.equals(unique.get(name)),
                    "PRIMARY".equalsIgnoreCase(name),
                    null,
                    null,
                    null,
                    null
                ));
            }
            return result;
        });
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return unchecked(() -> {
            DatabaseMetaData meta = requireConnection().getMetaData();
            List<ForeignKeyInfo> result = new ArrayList<>();
            try (ResultSet rs = meta.getImportedKeys(null, blankToNull(schema), table)) {
                while (rs.next()) {
                    String name = rs.getString("FK_NAME");
                    result.add(new ForeignKeyInfo(
                        name == null ? "" : name,
                        rs.getString("FKCOLUMN_NAME"),
                        rs.getString("PKTABLE_NAME"),
                        rs.getString("PKCOLUMN_NAME")
                    ));
                }
            }
            return result;
        });
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public String getTableDdl(String schema, String table) {
        List<IndexInfo> indexes;
        try {
            indexes = listIndexes(schema, table);
        } catch (RuntimeException e) {
            indexes = Collections.emptyList();
        }

        List<ForeignKeyInfo> foreignKeys;
        try {
            foreignKeys = listForeignKeys(schema, table);
        } catch (RuntimeException e) {
            foreignKeys = Collections.emptyList();
        }

        return DatabaseAgentKt.buildTableDdl(schema, table, getColumns(schema, table), indexes, foreignKeys);
    }

    @Override
    public QueryResult executeQuery(String sql, String schema, ExecuteQueryOptions options) {
        return JdbcExecutor.INSTANCE.execute(
            requireConnection(),
            sql,
            schema,
            this::setSchemaSQL,
            options.getMaxRows(),
            options.getFetchSize(),
            this::resultValue
        );
    }

    @Override
    public QueryPageResult executeQueryPage(String sql, String schema, QueryPageOptions options) {
        return JdbcExecutor.INSTANCE.executePage(
            requireConnection(),
            sql,
            schema,
            this::setSchemaSQL,
            options,
            this::resultValue
        );
    }

    @Override
    public QueryPageResult fetchQueryPage(String sessionId, int pageSize) {
        return JdbcExecutor.INSTANCE.fetchPage(sessionId, pageSize);
    }

    @Override
    public boolean closeQuerySession(String sessionId) {
        return JdbcExecutor.INSTANCE.closeQuerySession(sessionId);
    }

    @Override
    public QueryResult executeTransaction(List<String> statements, String schema) {
        return unchecked(() -> {
            Connection conn = requireConnection();
            boolean savedAutoCommit = conn.getAutoCommit();
            conn.setAutoCommit(false);
            long start = System.currentTimeMillis();
            try {
                if (schema != null && !schema.trim().isEmpty()) {
                    try (java.sql.Statement stmt = conn.createStatement()) {
                        stmt.execute(setSchemaSQL(schema));
                    }
                }

                long totalAffected = 0;
                for (String statement : statements) {
                    try (java.sql.Statement stmt = conn.createStatement()) {
                        totalAffected += stmt.executeUpdate(trimSql(statement));
                    }
                }
                conn.commit();
                return new QueryResult(
                    Collections.emptyList(),
                    Collections.emptyList(),
                    totalAffected,
                    System.currentTimeMillis() - start,
                    false
                );
            } catch (Exception e) {
                conn.rollback();
                throw e;
            } finally {
                conn.setAutoCommit(savedAutoCommit);
            }
        });
    }

    @Override
    public String setSchemaSQL(String schema) {
        if (profile.getSkipExecutionContext()) {
            return "";
        }
        return "SET SCHEMA " + JdbcIdentifiers.INSTANCE.doubleQuote(schema);
    }

    @Override
    public void disconnect() {
        uncheckedVoid(() -> {
            if (connection != null) {
                connection.close();
            }
            connection = null;
        });
    }

    protected Connection requireConnection() {
        if (connection == null) {
            throw new IllegalStateException("Not connected");
        }
        return connection;
    }

    protected Object resultValue(ResultSet rs, int index, int sqlType) {
        return unchecked(() -> {
            Object value;
            switch (sqlType) {
                case Types.BIGINT:
                    value = rs.getLong(index);
                    break;
                case Types.INTEGER:
                case Types.SMALLINT:
                case Types.TINYINT:
                    value = rs.getInt(index);
                    break;
                case Types.FLOAT:
                case Types.REAL:
                    value = rs.getFloat(index);
                    break;
                case Types.DOUBLE:
                    value = rs.getDouble(index);
                    break;
                case Types.DECIMAL:
                case Types.NUMERIC:
                    value = rs.getBigDecimal(index);
                    break;
                case Types.BOOLEAN:
                case Types.BIT:
                    value = rs.getBoolean(index);
                    break;
                default:
                    value = rs.getString(index);
                    break;
            }
            return rs.wasNull() ? null : value;
        });
    }

    private static String normalizeTableType(String type) {
        if (type == null || type.trim().isEmpty()) {
            return "TABLE";
        }
        if ("BASE TABLE".equals(type.toUpperCase(Locale.ROOT))) {
            return "TABLE";
        }
        return type;
    }

    private static Integer intOrNull(ResultSet rs, String column) throws Exception {
        Object value = rs.getObject(column);
        return value instanceof Number ? ((Number) value).intValue() : null;
    }

    private static Integer characterLength(ResultSet rs) throws Exception {
        String typeName = rs.getString("TYPE_NAME");
        if (typeName == null) {
            return null;
        }
        String normalized = typeName.toLowerCase(Locale.ROOT);
        if (!normalized.contains("char") && !normalized.contains("text")) {
            return null;
        }
        return intOrNull(rs, "COLUMN_SIZE");
    }

    private static String blankToNull(String value) {
        return value == null || value.trim().isEmpty() ? null : value;
    }

    private static void addNonBlank(Set<String> values, String value) {
        if (value != null && !value.trim().isEmpty()) {
            values.add(value);
        }
    }

    private static String trimSql(String sql) {
        String trimmed = sql.trim();
        while (trimmed.endsWith(";")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1).trim();
        }
        return trimmed;
    }

    private static <T> T unchecked(ThrowingSupplier<T> supplier) {
        try {
            return supplier.get();
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void uncheckedVoid(ThrowingRunnable runnable) {
        unchecked(() -> {
            runnable.run();
            return null;
        });
    }

    private static final class IndexColumn {
        private final int ordinal;
        private final String name;

        private IndexColumn(int ordinal, String name) {
            this.ordinal = ordinal;
            this.name = name;
        }

        private int getOrdinal() {
            return ordinal;
        }

        private String getName() {
            return name;
        }
    }

    private interface ThrowingSupplier<T> {
        T get() throws Exception;
    }

    private interface ThrowingRunnable {
        void run() throws Exception;
    }
}

File: common/src/main/kotlin/com/dbx/agent/ConfiguredJdbcAgent.kt
package com.dbx.agent

import java.sql.Connection
import java.sql.DatabaseMetaData
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.Types

data class JdbcAgentProfile(
    val driverClass: String,
    val urlTemplate: String,
    val defaultPort: Int = 0,
    val skipExecutionContext: Boolean = false,
    val excludedSchemas: Set<String> = emptySet(),
    val tableTypes: List<String> = listOf("TABLE", "VIEW", "MATERIALIZED VIEW", "SYSTEM TABLE", "SYSTEM VIEW"),
) {
    fun buildUrl(params: ConnectParams): String {
        if (params.connection_string.isNotBlank()) {
            return params.connection_string
        }
        val port = if (params.port > 0) params.port else defaultPort
        val base = urlTemplate
            .replace("{host}", params.host)
            .replace("{port}", port.toString())
            .replace("{database}", params.database)
        return appendUrlParams(base, params.url_params)
    }
}

private fun appendUrlParams(url: String, urlParams: String): String {
    val params = urlParams.trim().trimStart('?', '&')
    if (params.isBlank()) return url
    val separator = if (url.contains("?")) "&" else "?"
    return "$url$separator$params"
}

abstract class ConfiguredJdbcAgent(
    open val profile: JdbcAgentProfile,
) : DatabaseAgent {
    private var connection: Connection? = null
    private var configuredDatabase: String = ""

    override fun getConnection(): Connection? = connection

    override fun connect(params: ConnectParams) {
        Class.forName(profile.driverClass)
        connection = DriverManager.getConnection(profile.buildUrl(params), params.username, params.password)
        configuredDatabase = params.database
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName(profile.driverClass)
        DriverManager.getConnection(profile.buildUrl(params), params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        val conn = requireConnection()
        val names = linkedSetOf<String>()
        runCatching {
            conn.metaData.catalogs.use { rs ->
                while (rs.next()) {
                    rs.getString("TABLE_CAT")?.takeIf { it.isNotBlank() }?.let(names::add)
                }
            }
        }
        configuredDatabase.takeIf { it.isNotBlank() }?.let(names::add)
        conn.catalog?.takeIf { it.isNotBlank() }?.let(names::add)
        return names.map(::DatabaseInfo)
    }

    override fun listSchemas(): List<String> {
        val conn = requireConnection()
        val names = linkedSetOf<String>()
        val meta = conn.metaData
        runCatching {
            meta.getSchemas(null, null).use { rs ->
                while (rs.next()) {
                    rs.getString("TABLE_SCHEM")?.takeIf { it.isNotBlank() }?.let(names::add)
                }
            }
        }.recoverCatching {
            meta.schemas.use { rs ->
                while (rs.next()) {
                    rs.getString("TABLE_SCHEM")?.takeIf { it.isNotBlank() }?.let(names::add)
                }
            }
        }
        conn.schema?.takeIf { it.isNotBlank() }?.let(names::add)
        return names
            .filterNot { it.uppercase() in profile.excludedSchemas }
            .sorted()
    }

    override fun listTables(schema: String): List<TableInfo> {
        val meta = requireConnection().metaData
        val result = mutableListOf<TableInfo>()
        appendTables(result, meta, null, schema)
        if (result.isEmpty() && configuredDatabase.isNotBlank()) {
            appendTables(result, meta, configuredDatabase, schema)
        }
        return result.sortedBy { it.name }
    }

    private fun appendTables(result: MutableList<TableInfo>, meta: DatabaseMetaData, catalog: String?, schema: String) {
        meta.getTables(catalog, schema.ifBlank { null }, "%", profile.tableTypes.toTypedArray()).use { rs ->
            while (rs.next()) {
                result.add(
                    TableInfo(
                        name = rs.getString("TABLE_NAME"),
                        table_type = normalizeTableType(rs.getString("TABLE_TYPE")),
                        comment = rs.getString("REMARKS"),
                    )
                )
            }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val meta = requireConnection().metaData
        val primaryKeys = primaryKeys(meta, null, schema, table)
        val result = mutableListOf<ColumnInfo>()
        appendColumns(result, meta, null, schema, table, primaryKeys)
        if (result.isEmpty() && configuredDatabase.isNotBlank()) {
            val fallbackPrimaryKeys = primaryKeys(meta, configuredDatabase, schema, table)
            appendColumns(result, meta, configuredDatabase, schema, table, fallbackPrimaryKeys)
        }
        return result
    }

    private fun primaryKeys(meta: DatabaseMetaData, catalog: String?, schema: String, table: String): Set<String> {
        val keys = mutableSetOf<String>()
        runCatching {
            meta.getPrimaryKeys(catalog, schema.ifBlank { null }, table).use { rs ->
                while (rs.next()) {
                    rs.getString("COLUMN_NAME")?.let(keys::add)
                }
            }
        }
        return keys
    }

    private fun appendColumns(
        result: MutableList<ColumnInfo>,
        meta: DatabaseMetaData,
        catalog: String?,
        schema: String,
        table: String,
        primaryKeys: Set<String>,
    ) {
        meta.getColumns(catalog, schema.ifBlank { null }, table, "%").use { rs ->
            while (rs.next()) {
                val name = rs.getString("COLUMN_NAME")
                result.add(
                    ColumnInfo(
                        name = name,
                        data_type = rs.getString("TYPE_NAME"),
                        is_nullable = rs.getInt("NULLABLE") != DatabaseMetaData.columnNoNulls,
                        column_default = rs.getString("COLUMN_DEF"),
                        is_primary_key = name in primaryKeys,
                        comment = rs.getString("REMARKS"),
                        numeric_precision = intOrNull(rs, "COLUMN_SIZE"),
                        numeric_scale = intOrNull(rs, "DECIMAL_DIGITS"),
                        character_maximum_length = characterLength(rs),
                    )
                )
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        val meta = requireConnection().metaData
        val indexes = linkedMapOf<String, MutableList<Pair<Short, String>>>()
        val unique = mutableMapOf<String, Boolean>()
        meta.getIndexInfo(null, schema.ifBlank { null }, table, false, false).use { rs ->
            while (rs.next()) {
                val name = rs.getString("INDEX_NAME") ?: continue
                val column = rs.getString("COLUMN_NAME") ?: continue
                indexes.getOrPut(name) { mutableListOf() }.add(rs.getShort("ORDINAL_POSITION") to column)
                unique[name] = !rs.getBoolean("NON_UNIQUE")
            }
        }
        return indexes.map { (name, columns) ->
            IndexInfo(
                name = name,
                columns = columns.sortedBy { it.first }.map { it.second },
                is_unique = unique[name] ?: false,
                is_primary = name.equals("PRIMARY", ignoreCase = true),
            )
        }
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        val meta = requireConnection().metaData
        val result = mutableListOf<ForeignKeyInfo>()
        meta.getImportedKeys(null, schema.ifBlank { null }, table).use { rs ->
            while (rs.next()) {
                result.add(
                    ForeignKeyInfo(
                        name = rs.getString("FK_NAME") ?: "",
                        column = rs.getString("FKCOLUMN_NAME"),
                        ref_table = rs.getString("PKTABLE_NAME"),
                        ref_column = rs.getString("PKCOLUMN_NAME"),
                    )
                )
            }
        }
        return result
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> = emptyList()

    override fun executeQuery(sql: String, schema: String?, options: ExecuteQueryOptions): QueryResult {
        return JdbcExecutor.execute(
            requireConnection(),
            sql,
            schema,
            ::setSchemaSQL,
            maxRows = options.maxRows,
            fetchSize = options.fetchSize,
            valueReader = ::resultValue,
        )
    }

    override fun setSchemaSQL(schema: String): String {
        if (profile.skipExecutionContext) return ""
        return "SET SCHEMA ${JdbcIdentifiers.doubleQuote(schema)}"
    }

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    protected fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }

    protected open fun resultValue(rs: ResultSet, index: Int, sqlType: Int): Any? {
        val value = when (sqlType) {
            Types.BIGINT -> rs.getLong(index)
            Types.INTEGER, Types.SMALLINT, Types.TINYINT -> rs.getInt(index)
            Types.FLOAT, Types.REAL -> rs.getFloat(index)
            Types.DOUBLE -> rs.getDouble(index)
            Types.DECIMAL, Types.NUMERIC -> rs.getBigDecimal(index)
            Types.BOOLEAN, Types.BIT -> rs.getBoolean(index)
            else -> rs.getString(index)
        }
        return if (rs.wasNull()) null else value
    }
}

private fun normalizeTableType(type: String?): String {
    return when (type?.uppercase()) {
        "BASE TABLE" -> "TABLE"
        null, "" -> "TABLE"
        else -> type
    }
}

private fun intOrNull(rs: ResultSet, column: String): Int? {
    val value = rs.getObject(column) ?: return null
    return (value as Number).toInt()
}

private fun characterLength(rs: ResultSet): Int? {
    val typeName = rs.getString("TYPE_NAME")?.lowercase() ?: return null
    if (!typeName.contains("char") && !typeName.contains("text")) return null
    return intOrNull(rs, "COLUMN_SIZE")
}
